### ISSUE TITLE
Cookies changed on redirect not applied on redirected request

### ISSUE THREAD

**Issue by ArturGaspar** [scrapy/scrapy#5021]
### Description

If an existing cookie is assigned a new value on a redirect, the change is not applied in the redirected request.

### Steps to Reproduce

1. Make a request that sets a cookie.
2. Make a request that sets the same cookie with a different value and does a redirect.

**Expected behavior:** Redirected request will have cookie with value from step 2.

**Actual behavior:** Redirected request will have cookie with value from step 1, not from step 2.

**Reproduces how often:** Always

### Versions

```
Scrapy       : 2.2.0
lxml         : 4.6.2.0
libxml2      : 2.9.10
cssselect    : 1.1.0
parsel       : 1.5.2
w3lib        : 1.22.0
Twisted      : 20.3.0
Python       : 3.8.7 (default, Dec 22 2020, 08:33:13) [GCC]
pyOpenSSL    : 20.0.1 (OpenSSL 1.1.1h  22 Sep 2020)
cryptography : 3.3.2
Platform     : Linux-5.10.16-1-default-x86_64-with-glibc2.2.5
```

### Additional context

The spider below (can be run with `scrapy runspider`) demonstrates the issue and includes a workaround for the redirect middleware.

```python
from http.cookies import BaseCookie
from urllib.parse import urlparse

from scrapy import Request, Spider
from scrapy.downloadermiddlewares.redirect import RedirectMiddleware
from scrapy.http import Response


class DownloadHandler(object):
    """

    A fake download handler that simulates the behaviour that triggers the
    issue.

    """

    def download_request(self, request, spider):
        response = Response(request.url)
        path = urlparse(request.url).path

        if path == "/setup":
            response.headers["Set-Cookie"] = [b"c1=a", b"c2=b"]
        elif path == "/start":
            response.status = 302
            # Change one cookie and set one new cookie on redirect.
            response.headers["Set-Cookie"] = [b"c2=B", b"c3=c"]
            response.headers["Location"] = [b"/check_cookies"]
        elif path == "/check_cookies":
            cookies = BaseCookie(request.headers["Cookie"].decode())

            # Not changed on redirect. This always works.
            assert cookies["c1"].value == "a"

            # Initially unset, but is set on redirect. This always works.
            assert cookies["c3"].value == "c"

            # Set before redirect and changed on redirect. This is broken.
            assert cookies["c2"].value == "B"

        return response


class CookieFixRedirectMiddleware(RedirectMiddleware):
    def _redirect(self, redirected, request, spider, reason):
        redirected = super()._redirect(redirected, request, spider, reason)
        # Remove the cookie header so that the cookies middleware recreates it
        # with the changed values.
        redirected.headers.pop('Cookie', None)
        return redirected


class TestSpider(Spider):
    name = "test-redirect-cookie"

    custom_settings = {
        'DOWNLOAD_HANDLERS': {
            'http': __name__ + '.DownloadHandler'
        },
        # Uncomment below to obtain correct behaviour.

        # 'DOWNLOADER_MIDDLEWARES': {
        #     __name__ + '.CookieFixRedirectMiddleware': 600,
        #     'scrapy.downloadermiddlewares.redirect.RedirectMiddleware': None
        # }
    }

    def start_requests(self):
        yield Request("http://example.com/setup",
                      callback=self.parse_setup)

    def parse_setup(self, response):
        yield Request(response.urljoin("/start"),
                      callback=self.parse_cookie_test)

    def parse_cookie_test(self, response):
        pass
```

**Comment by ArturGaspar** [comment]
This is similar to #2141 in that it is caused by the redirect middleware not knowing whether a request value should be preserved.

**Comment by wRAR** [comment]
Hi Artur, this is #4717 and it was fixed in 2.4.0, see https://docs.scrapy.org/en/latest/news.html#backward-incompatible-changes
